static/app/api/apiOptions.spec.ts

```diff
@@ -0,0 +1,164 @@
+            releaseVersion: 'v1.0.0',
+          },
+        }
+      );
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<Release>>();
+    });
+
+    test('should not allow invalid path parameters', () => {
+      const options = apiOptions('/projects/$orgSlug/', {
+        staleTime: 0,
+        // @ts-expect-error Invalid path parameter
+        path: {orgSlug: 'my-org', invalidParam: 'invalid'},
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should not allow excess path parameters', () => {
+      const options = apiOptions('/projects/$orgSlug/', {
+        staleTime: 0,
+        // @ts-expect-error Excess path parameter
+        path: {orgSlug: 'my-org', extraParam: 'extra'},
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should require path params for paths with parameters', () => {
+      expect(() => {
+        const options = apiOptions('/projects/$orgSlug/', {
+          staleTime: 0,
+          // @ts-expect-error Missing required path parameter
+          path: {},
+        });
+
+        expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+      }).toThrow('Missing path param: orgSlug');
+    });
+
+    test('should not allow empty path parameters for paths without parameters', () => {
+      const options = apiOptions('/projects/', {
+        staleTime: 0,
+        // @ts-expect-error Empty path parameters not allowed
+        path: {},
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should not need path params for paths without parameters', () => {
+      const options = apiOptions('/projects/', {
+        staleTime: 0,
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should allow string or number path parameters', () => {
+      const options = apiOptions('/items/$id/', {
+        staleTime: 0,
+        path: {id: 123}, // number
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+
+      const options2 = apiOptions('/items/$id/', {
+        staleTime: 0,
+        path: {id: 'abc'}, // string
+      });
+
+      expectTypeOf(options2.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should default to never for unknown API paths', () => {
+      const options = apiOptions('/unknown/$param/', {
+        staleTime: 0,
+        path: {param: 'value'},
+      });
+
+      expectTypeOf(options.queryFn).returns.toEqualTypeOf<QueryFunctionResult<never>>();
+    });
+
+    test('should allow providing manual data type', () => {
+      const options = apiOptions.ReturnType<number>()('/foo/$bar', {
+        staleTime: 0,
+        path: {bar: 'baz'},
+      });
```

Comment:
This would be the recommended way to say: this endpoint will return `ApiResult<number>`. The naming is arbitrary, I’ve been thinking that `apiOptions.as<number>()` would also be nice. What matters is that it’s a function call and the type parameter is required and can’t be inferred.

It has to be split from the other type parameters to not mess up their type inference, as TypeScript still doesn’t have partial type inference.